python/tvm/relax/transform/transform.py

```diff
@@ -463,6 +463,16 @@ def KillAfterLastUse() -> tvm.ir.transform.Pass:
     return _ffi_api.KillAfterLastUse()  # type: ignore
 
 
+def ComputePrimValue() -> tvm.ir.transform.Pass:
+    """Compute all R.prim_value instances
```

Comment:
I think this description should be more precise. I assume it's supposed to come late in the phase ordering since it inserts direct calls to PrimFuncs? (And so should probably come after we end purity checking?)

Reply:
Good point on improving the docstring.

Regarding phase ordering, I don’t think we need to restrict its usage.  The calls to PrimFunc instances are valid in user-provided Relax functions, so this could occur early in the phase ordering.  The only limitation is that it must occur before `VMShapeLower`, as `VMShapeLower` expects all `R.prim_value(arg)` expressions to have int64 arguments.

Reply:
Yes, but inserting the PrimFunc calls will likely change the purity of the functions where that happens. `call_tir` could be used to avoid that but then that will require using this before lowering `call_tir`.

Reply:
Hmm.  I think this is another argument in favor of allowing `FuncStructInfo` annotations for `PrimFunc` objects, as that would allow the generated `PrimFunc` instances to be marked as pure functions.  I'll add a unit test to see how well that works for maintaining purity tracking when calling a pure PrimFunc from a pure Relax function.

Reply:
Are they truly pure? No modification of external values?

Reply:
I was thinking an even simpler heuristic: All PrimFuncs are impure, unless explicitly annotated otherwise.  In this case, since the functions are being generated in a manner that requires purity, it could also provide the annotation.

For long-term, agreed, it would be good to have the TIR-level purity analysis.  I think I'd weaken the condition you mentioned slightly: A function is impure if it writes to a buffer that it didn't itself allocate.

Reply:
Good point. I'm fine with requiring an annotation since the great majority of PrimFuncs are going to be impure.